File: frontend/src/lib/api/cv.ts
/**
 * CV API client
 */

import { apiClient } from "./client";

export type ParsingStatus = "pending" | "processing" | "completed" | "failed";

export interface Experience {
  title: string;
  company: string;
  location: string | null;
  start_date: string | null;
  end_date: string | null;
  description: string;
  highlights: string[];
}

export interface Education {
  degree: string;
  institution: string;
  location: string | null;
  graduation_date: string | null;
  gpa: string | null;
}

export interface ParsedCV {
  full_name: string;
  email: string | null;
  phone: string | null;
  location: string | null;
  summary: string | null;
  skills: string[];
  experience: Experience[];
  education: Education[];
  certifications: string[];
  languages: string[];
}

export interface CV {
  id: string;
  user_id: string;
  filename: string;
  file_path: string;
  raw_text: string | null;
  parsed_data: ParsedCV | null;
  parsing_status: ParsingStatus;
  parsing_error: string | null;
  is_primary: boolean;
  created_at: string;
  updated_at: string;
}

export interface CVListItem {
  id: string;
  filename: string;
  is_primary: boolean;
  parsing_status: ParsingStatus;
  created_at: string;
  updated_at: string;
}

export interface CVUploadResponse {
  id: string;
  filename: string;
  file_path: string;
  message: string;
}

export interface CVParseResponse {
  id: string;
  parsing_status: ParsingStatus;
  parsing_error: string | null;
  parsed_data: ParsedCV | null;
  message: string;
}

export const cvApi = {
  /**
   * Upload a new CV
   */
  async upload(file: File): Promise<CVUploadResponse> {
    const formData = new FormData();
    formData.append("file", file);

    const response = await apiClient.post<CVUploadResponse>(
      "/cv/upload",
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );

    return response.data;
  },

  /**
   * List all CVs for the current user
   */
  async list(): Promise<CVListItem[]> {
    const response = await apiClient.get<CVListItem[]>("/cv/");
    return response.data;
  },

  /**
   * Get CV details by ID
   */
  async get(cvId: string): Promise<CV> {
    const response = await apiClient.get<CV>(`/cv/${cvId}`);
    return response.data;
  },

  /**
   * Delete a CV
   */
  async delete(cvId: string): Promise<void> {
    await apiClient.delete(`/cv/${cvId}`);
  },

  /**
   * Set a CV as primary
   */
  async setPrimary(cvId: string): Promise<CV> {
    const response = await apiClient.patch<CV>(`/cv/${cvId}/primary`);
    return response.data;
  },

  /**
   * Download a CV file
   */
  async download(cvId: string, filename: string): Promise<void> {
    const response = await apiClient.get(`/cv/${cvId}/download`, {
      responseType: "blob",
    });

    // Create a download link
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement("a");
    link.href = url;
    link.setAttribute("download", filename);
    document.body.appendChild(link);
    link.click();
    link.remove();
    window.URL.revokeObjectURL(url);
  },

  /**
   * Get a blob URL for PDF preview
   */
  async getPreviewUrl(cvId: string): Promise<string> {
    const response = await apiClient.get(`/cv/${cvId}/download`, {
      responseType: "blob",
    });

    return window.URL.createObjectURL(
      new Blob([response.data], { type: "application/pdf" })
    );
  },

  /**
   * Trigger CV parsing (PDF text extraction + LLM structuring)
   */
  async parse(cvId: string): Promise<CVParseResponse> {
    const response = await apiClient.post<CVParseResponse>(`/cv/${cvId}/parse`);
    return response.data;
  },

  /**
   * Manually update parsed CV data
   */
  async updateParsedData(cvId: string, data: Partial<ParsedCV>): Promise<CV> {
    const response = await apiClient.patch<CV>(`/cv/${cvId}/parsed-data`, data);
    return response.data;
  },
};

File: backend/app/schemas/cv.py
"""CV schemas for request/response validation."""

import enum
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ParsingStatusSchema(str, enum.Enum):
    """CV parsing status enum for API schemas."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class CVBase(BaseModel):
    """Base CV schema with common fields."""

    filename: str = Field(..., max_length=255, description="Original filename")
    is_primary: bool = Field(default=False, description="Whether this is the primary CV")


class CVCreate(CVBase):
    """Schema for creating a new CV (internal use)."""

    file_path: str = Field(..., max_length=512, description="Storage path")
    raw_text: str | None = Field(None, description="Extracted text from PDF")
    parsed_data: dict[str, Any] | None = Field(None, description="Parsed CV data")


class CVUpdate(BaseModel):
    """Schema for updating CV metadata."""

    is_primary: bool | None = Field(None, description="Set as primary CV")


class CVResponse(CVBase):
    """Schema for CV response."""

    id: uuid.UUID
    user_id: uuid.UUID
    file_path: str
    raw_text: str | None
    parsed_data: dict[str, Any] | None
    parsing_status: ParsingStatusSchema
    parsing_error: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CVListResponse(BaseModel):
    """Schema for listing CVs."""

    id: uuid.UUID
    filename: str
    is_primary: bool
    parsing_status: ParsingStatusSchema
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CVUploadResponse(BaseModel):
    """Schema for CV upload response."""

    id: uuid.UUID
    filename: str
    file_path: str
    message: str = "CV uploaded successfully"

    model_config = ConfigDict(from_attributes=True)


# --- Parsed CV Data Schemas ---


class Experience(BaseModel):
    """Work experience entry from parsed CV."""

    title: str = Field(..., description="Job title")
    company: str = Field(..., description="Company name")
    location: str | None = Field(None, description="Job location")
    start_date: str | None = Field(None, description="Start date")
    end_date: str | None = Field(None, description="End date or 'Present'")
    description: str = Field(default="", description="Role description")
    highlights: list[str] = Field(default_factory=list, description="Key achievements")


class Education(BaseModel):
    """Education entry from parsed CV."""

    degree: str = Field(..., description="Degree/qualification name")
    institution: str = Field(..., description="School/university name")
    location: str | None = Field(None, description="Institution location")
    graduation_date: str | None = Field(None, description="Graduation date")
    gpa: str | None = Field(None, description="GPA if available")


class ParsedCV(BaseModel):
    """Structured data extracted from CV using LLM."""

    full_name: str = Field(..., description="Full name of the candidate")
    email: str | None = Field(None, description="Email address")
    phone: str | None = Field(None, description="Phone number")
    location: str | None = Field(None, description="Current location")
    summary: str | None = Field(None, description="Professional summary/objective")
    skills: list[str] = Field(default_factory=list, description="List of skills")
    experience: list[Experience] = Field(default_factory=list, description="Work experience")
    education: list[Education] = Field(default_factory=list, description="Education history")
    certifications: list[str] = Field(default_factory=list, description="Certifications")
    languages: list[str] = Field(default_factory=list, description="Languages spoken")


class CVParseResponse(BaseModel):
    """Response for CV parsing endpoint."""

    id: uuid.UUID
    parsing_status: ParsingStatusSchema
    parsing_error: str | None = None
    parsed_data: ParsedCV | None = None
    message: str = "Parsing initiated"

    model_config = ConfigDict(from_attributes=True)


class CVParsedDataUpdate(BaseModel):
    """Schema for manually updating parsed CV data."""

    full_name: str | None = Field(None, description="Full name of the candidate")
    email: str | None = Field(None, description="Email address")
    phone: str | None = Field(None, description="Phone number")
    location: str | None = Field(None, description="Current location")
    summary: str | None = Field(None, description="Professional summary/objective")
    skills: list[str] | None = Field(None, description="List of skills")
    experience: list[Experience] | None = Field(None, description="Work experience")
    education: list[Education] | None = Field(None, description="Education history")
    certifications: list[str] | None = Field(None, description="Certifications")
    languages: list[str] | None = Field(None, description="Languages spoken")
